Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55064 Replace torch.chain_matmul calls to torch.linalg.multi_dot
* **#53453 Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot**
* #52859 [Relanding] Implemented torch.linalg.multi_dot

Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

heitorschueroff committed:
Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

mruberry commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -281,7 +281,7 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
   KERNEL(ADD_NS(baddbmm), "baddbmm", Tensor (const Tensor &, const Tensor &, const Tensor &, Scalar, Scalar), fp16)
   KERNEL(ADD_NS(bmm), "bmm", Tensor (const Tensor &, const Tensor &), fp16)
-  KERNEL(ADD_NS(chain_matmul), "chain_matmul", Tensor (TensorList), fp16)
+  KERNEL(ADD_NS(linalg_multi_dot), "linalg_multi_dot", Tensor (TensorList), fp16)
```
We can keep chain_matmul() here, too, right?

mruberry commented on test/test_autograd.py:
```diff
@@ -2936,20 +2936,6 @@ def test_igammac(self):
         gradcheck(torch.igamma, (s, x))
         gradgradcheck(torch.igamma, (s, x))
 
-    def test_chain_matmul(self):
```
Let's also change calls to chain_matmul in the code base to linalg.multi_dot, like here:

https://github.com/pytorch/pytorch/blob/7297556d5d9561a1c0a4b96ad068c13e05ff9a5c/test/test_autograd.py

and 

https://github.com/pytorch/pytorch/blob/6dc763df3098485f158eeed2b9621cffaa229669/torch/nn/utils/spectral_norm.py

mruberry commented on test/test_linalg.py:
```diff
@@ -3274,8 +3274,21 @@ def test_old_matrix_rank(self, device, dtype):
-    @dtypes(torch.float, torch.cfloat)
-    @precisionOverride({torch.float: 1e-02, torch.cfloat: 1e-02})
+    @dtypes(torch.double)
+    def test_chain_matmul(self, device, dtype):
```
Add a comment explaining that this test is just checking for chain_matmul()'s additional error checking vs. linalg.multi_dot()

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -390,18 +390,22 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
-        [(2, S), (S, 2), (2, S)],
+    # Each test case consists of the sizes in the chain of multiplications
+    # e.g. [2, S, 2, S] generates matrices (2, S) @ (S, 2) @ (2, S)
+    test_cases = [
```
How much does this increase test time by?

mruberry commented on torch/functional.py:
```diff
@@ -1438,6 +1438,10 @@ def chain_matmul(*matrices):
     If :math:`N` is 1, then this is a no-op - the original matrix is returned as is.
 
+    .. note::
+        :func:`torch.chain_matmul` is deprecated, use :func:`torch.linalg.multi_dot` instead. The differences are
```
Let's make this a warning and not explain the differences. The one case where a user of chain_matmul() would be frustrated trying to switch to multi_dot() is if they're calling it with a single tensor, which seems incredibly unlikely and is easy to discover.

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

mruberry commented on torch/functional.py:
```diff
@@ -1470,17 +1470,19 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
 
-def chain_matmul(*matrices):
+def chain_matmul(*matrices, out=None):
     r"""Returns the matrix product of the :math:`N` 2-D tensors. This product is efficiently computed
```
This is OK and we don't need to block the PR on this, but I think this can just say that chain_matmul is an alias for multi_dot?

mruberry reviewed (approved):
Cool! Let's get this merged. We can worry about docs refinements in our docs review before the 1.9 release.

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"




[ghstack-poisoned]

anjali411 commented on test/test_linalg.py:
```diff
@@ -3658,6 +3658,21 @@ def test_old_matrix_rank(self, device, dtype):
             self.assertEqual(torch.matrix_rank(aaT, 0.01, True), np.linalg.matrix_rank(aaT.cpu().numpy(), 0.01, True))
 
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.double)
```
```suggestion
    @dtypes(torch.double, torch.cdouble)
```

heitorschueroff committed:
Update test/test_linalg.py

Co-authored-by: anjali411 <chourdiaanjali123@gmail.com>

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

ezyang commented:
ACKing `native_functions.yaml` changes

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
added wrapper function to pack inputs to multi_dot in distributed testing on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot"


Differential Revision: [D27406282](https://our.internmc.facebook.com/intern/diff/D27406282)

[ghstack-poisoned]
